src/storage/gc_worker.rs

```diff
@@ -0,0 +1,279 @@
+            engine,
+            ratio_threshold,
+
+            stats: StatisticsSummary::default(),
+            empty_range_counter: KV_GC_EMPTY_RANGE_COUNTER.local(),
+            skipped_counter: KV_GC_SKIPPED_COUNTER.local(),
+            duration_histogram: GC_DURATION_HISTOGRAM.local(),
+            gctask_counter: GC_GCTASK_COUNTER.local(),
+            gctask_fail_counter: GC_GCTASK_FAIL_COUNTER.local(),
+            gc_keys_counter: GC_KEYS_COUNTER_VEC.local(),
+        }
+    }
+
+    // Scan keys in the region. Returns scanned keys if any, and a key indicating scan progress
+    fn scan_keys(
+        &mut self,
+        ctx: &Context,
+        safe_point: u64,
+        from: Option<Key>,
+        limit: usize,
+    ) -> Result<(Option<Vec<Key>>, Option<Key>)> {
+        let is_range_start_gc = from.is_none();
+
+        let snapshot = self.engine.snapshot(ctx)?;
+        let mut reader = MvccReader::new(
+            snapshot,
+            Some(ScanMode::Forward),
+            !ctx.get_not_fill_cache(),
+            None,
+            None,
+            ctx.get_isolation_level(),
+        );
+
+        let need_gc = (!is_range_start_gc) || reader.need_gc(safe_point, self.ratio_threshold);
+        let res = if !need_gc {
+            self.skipped_counter.inc();
+            Ok((None, None))
+        } else {
+            reader
+                .scan_keys(from, limit)
+                .map_err(Error::from)
+                .and_then(|(keys, next)| {
+                    if keys.is_empty() {
+                        if is_range_start_gc {
+                            self.empty_range_counter.inc();
+                        }
+                        Ok((None, None))
+                    } else {
+                        Ok((Some(keys), next))
+                    }
+                })
+        };
+        self.stats.add_statistics(reader.get_statistics());
+        res
+    }
+
+    // Clean up outdated data.
+    fn gc_keys(
+        &mut self,
+        ctx: &Context,
+        safe_point: u64,
+        keys: Vec<Key>,
+        mut next_scan_key: Option<Key>,
+    ) -> Result<Option<Key>> {
+        let snapshot = self.engine.snapshot(ctx)?;
+        let mut txn = MvccTxn::new(
+            snapshot,
+            0,
+            Some(ScanMode::Forward),
+            ctx.get_isolation_level(),
+            !ctx.get_not_fill_cache(),
+        );
+        for k in keys {
+            txn.gc(&k, safe_point)?;
+            if txn.write_size() >= MAX_TXN_WRITE_SIZE {
+                next_scan_key = Some(k);
+                break;
+            }
+        }
+        self.stats.add_statistics(txn.get_statistics());
+
+        let modifies = txn.into_modifies();
+        if !modifies.is_empty() {
+            self.engine.write(ctx, modifies)?;
+        }
+        Ok(next_scan_key)
+    }
+
+    pub fn gc(&mut self, ctx: Context, safe_point: u64) -> Result<()> {
+        let _gc_timer = self.duration_histogram.start_coarse_timer();
```

Comment:
This seems to have no purpose?

Reply:
It will record the duration from here to where it is dropped